runtime/lua/vim/lsp.lua

```diff
@@ -811,78 +811,117 @@ local function once(fn)
 --- Notify all attached clients that a buffer has changed.
 local text_document_did_change_handler
 do
-  local encoding_index = { ["utf-8"] = 1; ["utf-16"] = 2; ["utf-32"] = 3; }
-  text_document_did_change_handler = function(_, bufnr, changedtick,
-      firstline, lastline, new_lastline, old_byte_size, old_utf32_size,
-      old_utf16_size)
+  -- on_bytes handler.
+  -- See nvim_buf_attach() -> on_bytes
+  text_document_did_change_handler = function(
+      _, bufnr, changedtick,
+      start_row, start_column, byte_offset,
+      old_end_row, old_end_column, old_end_byte_length,
+      new_end_row, new_end_column, new_end_byte_length)
+
+    -- TODO: New logging
 
     local _ = log.debug() and log.debug(
-      string.format("on_lines bufnr: %s, changedtick: %s, firstline: %s, lastline: %s, new_lastline: %s, old_byte_size: %s, old_utf32_size: %s, old_utf16_size: %s",
-      bufnr, changedtick, firstline, lastline, new_lastline, old_byte_size, old_utf32_size, old_utf16_size),
-      nvim_buf_get_lines(bufnr, firstline, new_lastline, true)
+      string.format("on_bytes bufnr: %s, changedtick: %s, start_row: %s, " ..
+                    "start_column: %s, byte_offset: %s, old_end_row: %s, " ..
+                    "old_end_column: %s, old_end_byte_length: %s, " ..
+                    "new_end_row: %s, new_end_column: %s, new_end_byte_length: %s",
+                    bufnr, changedtick, start_row, start_column, byte_offset,
+                    old_end_row, old_end_column, old_end_byte_length, new_end_row,
+                    new_end_column, new_end_byte_length),
+      buf_range_to_text(bufnr, {
+          start_row = start_row,
+          start_column = start_column,
+          new_end_row=new_end_row,
+          new_end_column = new_end_column,
+          })
     )
 
     -- Don't do anything if there are no clients attached.
     if tbl_isempty(all_buffer_active_clients[bufnr] or {}) then
       return
     end
 
-    util.buf_versions[bufnr] = changedtick
-    -- Lazy initialize these because clients may not even need them.
-    local incremental_changes = once(function(client)
-      local size_index = encoding_index[client.offset_encoding]
-      local length = select(size_index, old_byte_size, old_utf16_size, old_utf32_size)
-      local lines = nvim_buf_get_lines(bufnr, firstline, new_lastline, true)
-
-      -- This is necessary because we are specifying the full line including the
-      -- newline in range. Therefore, we must replace the newline as well.
-      if #lines > 0 then
-       table.insert(lines, '')
+    local incremental_changes = once(function(_client)
+      local text = buf_range_to_text(bufnr, {
+                      start_row = start_row,
+                      start_column = start_column,
+                      new_end_row=new_end_row,
+                      new_end_column = new_end_column,
+                      })
+
+      if new_end_byte_length < old_end_byte_length then
```

Comment:
This is not correct. `text` would be empty only when `new_end_byte_length` is zero, but if `buf_range_to_text` is correct that will already be the case.

Reply:
Fixed, thank you

Reply:
Wait, I think we actually need this. I'm getting errors now when doing live editing.

Reply:
Well you can check if `new_end_row` and `new_end_column` is both zero in `buf_range_to_text` and then return empty string even before API-call.